From d7cdf1e47c596c3fa6cffedd4904c67919389668 Mon Sep 17 00:00:00 2001
From: "Todd C. Miller" <Todd.Miller@sudo.ws>
Date: Tue, 14 Sep 2021 08:02:37 -0600
Subject: [PATCH] append_defaults() should not be passed a value for boolean
 flags. The operation should simply be set to true/false. Also treat a NULL
 file as coming from the front-end. Bug #993.

---
 plugins/sudoers/defaults.c | 3 +++
 plugins/sudoers/policy.c   | 4 ++--
 2 files changed, 5 insertions(+), 2 deletions(-)

diff --git a/plugins/sudoers/defaults.c b/plugins/sudoers/defaults.c
index 5739b40e2..95e6fbd30 100644
--- a/plugins/sudoers/defaults.c
+++ b/plugins/sudoers/defaults.c
@@ -206,6 +206,9 @@ parse_default_entry(struct sudo_defs_types *def, const char *val, int op,
     int rc;
     debug_decl(parse_default_entry, SUDOERS_DEBUG_DEFAULTS);
 
+    if (file == NULL)
+	file = "front-end";
+
     sudo_debug_printf(SUDO_DEBUG_INFO, "%s: %s:%d:%d: %s=%s op=%d",
 	__func__, file, line, column, def->name, val ? val : "", op);
 
diff --git a/plugins/sudoers/policy.c b/plugins/sudoers/policy.c
index eb1ce43b7..c8c20735d 100644
--- a/plugins/sudoers/policy.c
+++ b/plugins/sudoers/policy.c
@@ -224,7 +224,7 @@ sudoers_policy_deserialize_info(void *v, struct defaults_list *defaults)
 	if (MATCHES(*cur, "prompt=")) {
 	    /* Allow epmpty prompt. */
 	    user_prompt = *cur + sizeof("prompt=") - 1;
-	    if (!append_default("passprompt_override", "true", true, NULL, defaults))
+	    if (!append_default("passprompt_override", NULL, true, NULL, defaults))
 		goto oom;
 	    continue;
 	}
@@ -285,7 +285,7 @@ sudoers_policy_deserialize_info(void *v, struct defaults_list *defaults)
 	if (MATCHES(*cur, "login_class=")) {
 	    CHECK(*cur, "login_class=");
 	    login_class = *cur + sizeof("login_class=") - 1;
-	    if (!append_default("use_loginclass", "true", true, NULL, defaults))
+	    if (!append_default("use_loginclass", NULL, true, NULL, defaults))
 		goto oom;
 	    continue;
 	}
